### PR TITLE
Revert "[cmake] Xcode 14 warns about -undefined dynamic_lookup; simply suppress:"

### LINKED ISSUE #11937: macOS linking `-flat_namespace` changes user-facing behavior
From @henryiii :

> ROOT can't pass homebrew's audit any more because it is now injecting `-flat_namespace`. Flat namespaces have been deprecated since the move to Intel. I believe this was done due to `-undefined dynamic_lookup` producing a warning - Apple plans to add something to replace this in the future, but it's not the old `-flat_namespace`. See [Homebrew/homebrew-core#114543 (comment)](https://github.com/Homebrew/homebrew-core/pull/114543#issuecomment-1330719444)
> Flat namespace makes major changes to the layout, is more deprecated than dynamic_lookup, and isn't supposed to even work in conda-forge. See also https://github.com/pybind/pybind11/pull/4301 .
> `-undefined dynamic_lookup` does silence the linker on missing symbols, that's what it's for. The problem is that the method it uses is deprecated, producing a warning (and already broken in iOS) `-flat_namespace` doesn't just silence the linker, it changes the binary layout for every symbol. You could always just silence the warning `-Wl,-w` or `-Wl,-no_fixup_chains`; this will eventually break, but is less deprecated than `-flat_namespace`.
> Apple's official statement is basically "live with the warning for now, we'll find a way to not break all CPython extensions eventually" from what I gather.

### PR DESCRIPTION
**Pull request by Axel-Naumann**
See #11937 for why `-flat_namespace` is bad.

This reverts commit a05d4be.

This PR fixes #11937

### CODE REVIEW

**Comment by pcanal** [comment]
Additional mention of the original issue (chained lookups): https://developer.apple.com/forums/thread/719961, https://github.com/python/cpython/issues/97524 
and https://openradar.appspot.com/radar?id=5536824084660224.

Instead of `-Wl,-w` (which suppress all warnings), an undocumented options seems to be `-Xlinker -no_fixup_chains` (humm I guess actually maybe `-Wl,-no_fixup_chains`)

File: bindings/jupyroot/CMakeLists.txt [hunk 1]
```diff
@@ -47,7 +47,7 @@ foreach(val RANGE ${how_many_pythons})
     set_target_properties(${libname} PROPERTIES SUFFIX ".pyd")
     target_link_libraries(${libname} PUBLIC Core  ${python_library})
   elseif(APPLE)
-    target_link_libraries(${libname} PUBLIC -Wl,-bind_at_load -Wl,-flat_namespace -Wl,-undefined -Wl,suppress Core)
+    target_link_libraries(${libname} PUBLIC -Wl,-bind_at_load -Wl,-undefined -Wl,dynamic_lookup Core)
```
Comment (pcanal) [inline bindings/jupyroot/CMakeLists.txt:50]:
```suggestion
    target_link_libraries(${libname} PUBLIC -Wl,-bind_at_load -Wl,-w -Wl,-undefined -Wl,dynamic_lookup Core)
```

File: bindings/pyroot/cppyy/CPyCppyy/CMakeLists.txt [hunk 2]
```diff
@@ -63,7 +63,7 @@ foreach(val RANGE ${how_many_pythons})
     set_target_properties(${libname} PROPERTIES PREFIX "lib")
     set_target_properties(${libname} PROPERTIES SUFFIX ".pyd")
   elseif(APPLE)
-    target_link_libraries(${libname} PUBLIC -Wl,-bind_at_load -Wl,-flat_namespace -Wl,-undefined -Wl,suppress cppyy_backend${python_under_version_string})
+    target_link_libraries(${libname} PUBLIC -Wl,-bind_at_load -Wl,-undefined -Wl,dynamic_lookup cppyy_backend${python_under_version_string})
```
Comment (pcanal) [inline bindings/pyroot/cppyy/CPyCppyy/CMakeLists.txt:66]:
```suggestion
    target_link_libraries(${libname} PUBLIC -Wl,-bind_at_load -Wl,-w -Wl,-undefined -Wl,dynamic_lookup cppyy_backend${python_under_version_string})
```

File: bindings/pyroot/pythonizations/CMakeLists.txt [hunk 3]
```diff
@@ -168,7 +168,7 @@ foreach(val RANGE ${how_many_pythons})
     set_target_properties(${libname} PROPERTIES SUFFIX ".pyd")
     target_link_libraries(${libname} PUBLIC Core Tree cppyy${python_under_version_string})
   elseif(APPLE)
-    target_link_libraries(${libname} PUBLIC -Wl,-bind_at_load -Wl,-flat_namespace -Wl,-undefined -Wl,suppress Core Tree cppyy${python_under_version_string})
+    target_link_libraries(${libname} PUBLIC -Wl,-bind_at_load -Wl,-undefined -Wl,dynamic_lookup Core Tree cppyy${python_under_version_string})
```
Comment (pcanal) [inline bindings/pyroot/pythonizations/CMakeLists.txt:171]:
```suggestion
    target_link_libraries(${libname} PUBLIC -Wl,-bind_at_load -Wl,-w -Wl,-undefined -Wl,dynamic_lookup Core Tree cppyy${python_under_version_string})
```

**Review by pcanal — APPROVED** [review]
With a few more `-Wl,-w` it should work (or maybe we could disable the chained fixups)

**Comment by Axel-Naumann** [comment]
> Instead of -Wl,-w (which suppress all warnings), an undocumented options seems to be -Xlinker -no_fixup_chains (humm I guess actually maybe -Wl,-no_fixup_chains)

I am afraid that older linkers will not know `-no_fixup_chains` as that seems to be a new concern. So `-w` it is.

**Comment by pcanal** [comment]
> I am afraid that older linkers will not know -no_fixup_chains as that seems to be a new concern. 

Did you try on the oldest (not that old :)) XCode we support?

**Comment by Axel-Naumann** [comment]
My `ld`:
```
$ ld -no_fixup_chains /dev/null -o /dev/null
ld: warning: platform not specified
ld: warning: -arch not specified
ld: warning: No platform min-version specified on command line
ld: file too small (length=0) file '/dev/null' for architecture unknown
```

On 10.15 (i.e. just dead):
```
$ ld -no_fixup_chains /dev/null -o /dev/null
ld: unknown option: -no_fixup_chains
```
q.e.d. :-)

**Comment by Axel-Naumann** [comment]
macOS 11:
```
% ld -no_fixup_chains /dev/null -o /dev/null
ld: unknown option: -no_fixup_chains
```

**Comment by Axel-Naumann** [comment]
The Windows "failure" seems spurious:
```
18:21:16  Build failed because:
[Pipeline] echo
18:21:16  hudson.AbortException: script returned exit code -1
```
even though all tests have passed.

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on ROOT-performance-centos8-multicore/cxx17

**Comment by Axel-Naumann** [comment]
v6-28: https://github.com/root-project/root/pull/12100

**Pull request merged**